### PR TITLE
Export cgmes zip

### PR DESCRIPTION
**Pull request by etiennehomer**
Signed-off-by: HOMER Etienne <etiennehomer@gmail.com>

### CODE REVIEW

**Review by FranckLecuyer — DISMISSED** [review]

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java [hunk 1]
```diff
@@ -45,21 +67,113 @@ public NetworkConversionService(RestTemplate restTemplate) {
...
+            List<FileInfos> mergedIgms = caseFetcherService.getCases(caseUuids);
+            for (FileInfos mergedIgm : mergedIgms) {
+                cgmesProfiles.addAll(unzipCgmes(mergedIgm));
+            }
+
+            //Add SV profile
+            cgmesProfiles.add(getSvProfile(networkUuids, baseFileName));
+
+            //Add boundary files
+            cgmesProfiles.addAll(getBoundaries());
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:85]: We suppose for this dev that we have only one pair of  boundaries files in the database.
That's what I understood. Ok with that @geofjamg  @ChristineHesse  ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java [hunk 2]
```diff
@@ -45,21 +67,113 @@ public NetworkConversionService(RestTemplate restTemplate) {
...
+                int length = -1;
+                long totalBytes = 0;
+                baos = new ByteArrayOutputStream();
+                while (totalBytes < MAX_ZIP_ENTRY_SIZE && (length = zis.read(buffer)) > 0) {
+                    baos.write(buffer, 0, length);
+                    totalBytes = totalBytes + length;
+                }
+                //Remove repertory name before file name
+                fileName = FilenameUtils.getName(entry.getName());
+                isEntryToAdd = !fileName.equals("") && !fileName.endsWith(EQ_BD_FILE_NAME_SUFFIXE) && !fileName.endsWith(TP_BD_FILE_NAME_SUFFIXE) && !fileName.matches(SV_PROFILE_REGEX);
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java]: I checked with @ChristineHesse . This pattern 20200629T0000Z__ENTSOE_EQBD_001.xml is correct. I changed the regex to fix that.

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java [hunk 3]
```diff
@@ -19,23 +21,43 @@
...
     private static final String DELIMITER = "/";
+    private static final String CGMES_FORMAT = "CGMES";
+    private static final String EQ_BD_FILE_NAME_SUFFIXE = "_EQ_BD.xml";
+    private static final String TP_BD_FILE_NAME_SUFFIXE = "_TP_BD.xml";
+    private static final String SV_PROFILE = "SV";
+    private static final String UNDERSCORE = "_";
+    private static final String BUSINESS_PROCESS = "1D";
+    private static final String FILE_VERSION = "1";
+    private static final String XML_EXTENSION = ".xml";
+    private static final String XML_ZIP = ".zip";
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java]: done

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:94]: Fixed. We get the last boundaries on cgmes-boundary-server

File: src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java [hunk 4]
```diff
@@ -107,6 +105,19 @@ private String getSearchQuery(List<String> tsos, ZonedDateTime dateTime, String
         return Collections.emptyList();
     }
 
+    public List<FileInfos> getCases(List<UUID> caseUuids) {
+        List<FileInfos> cases = new ArrayList<>();
+        UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromPath(DELIMITER + CASE_API_VERSION + "/cases/{caseUuid}");
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java:110]: Don't use UriComponentsBuilder directly like this, it's unsafe. (in this case it works, but if there were weird characters in `(DELIMITER + CASE_API_VERSION + "/cases/{caseUuid}` or `queryParam("xiidm", false);` then it would break.

Instead, pass everything to resttemplate.exchange():
```java
String uriTemplate = DELIMITER + CASE_API_VERSION + "/cases/{caseUuid}?xiidm=false";
caseServerRest.exchange(uriTemplate, ..., caseUuid.toString())
```

File: src/main/java/org/gridsuite/merge/orchestrator/server/CgmesBoundaryService.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        this.cgmesBoundaryServerRest = builder.uriTemplateHandler(new DefaultUriBuilderFactory(cgmesBoundaryServerBaseUri))
+                .build();
+    }
+
+    public CgmesBoundaryService(RestTemplate restTemplate) {
+        this.cgmesBoundaryServerRest = restTemplate;
+    }
+
+    public List<BoundaryInfos> getBoundaries() {
+        String uri = UriComponentsBuilder.fromPath(DELIMITER + CGMES_BOUNDARY_API_VERSION + "/boundaries/last").build().toUriString();
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/CgmesBoundaryService.java:60]: same, don't use UriComponentsBuilder here, give the string templte directly to restTemplate

File: src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java [hunk 6]
```diff
@@ -107,6 +105,19 @@ private String getSearchQuery(List<String> tsos, ZonedDateTime dateTime, String
...
+        uriBuilder = uriBuilder.queryParam("xiidm", false);
+        String uri = uriBuilder.build().toUriString();
+        for (UUID caseUuid : caseUuids) {
+            ResponseEntity<byte[]> responseEntity = caseServerRest.exchange(uri, HttpMethod.GET, HttpEntity.EMPTY, new ParameterizedTypeReference<byte[]>() { }, caseUuid.toString());
+            String fileName = caseUuid.toString();
+            cases.add(new FileInfos(fileName.concat(".zip"), responseEntity.getBody()));
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java:116]: why add .zip manually here ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java [hunk 7]
```diff
@@ -107,6 +105,19 @@ private String getSearchQuery(List<String> tsos, ZonedDateTime dateTime, String
         return Collections.emptyList();
     }
 
+    public List<FileInfos> getCases(List<UUID> caseUuids) {
+        List<FileInfos> cases = new ArrayList<>();
+        UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromPath(DELIMITER + CASE_API_VERSION + "/cases/{caseUuid}");
+        uriBuilder = uriBuilder.queryParam("xiidm", false);
+        String uri = uriBuilder.build().toUriString();
+        for (UUID caseUuid : caseUuids) {
+            ResponseEntity<byte[]> responseEntity = caseServerRest.exchange(uri, HttpMethod.GET, HttpEntity.EMPTY, new ParameterizedTypeReference<byte[]>() { }, caseUuid.toString());
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java:114]: Is it ok to load all the cases in memory, each in a byte[]?

File: src/main/java/org/gridsuite/merge/orchestrator/server/MergeEventService.java [hunk 8]
```diff
@@ -51,11 +51,11 @@ public MergeEventService(MergeRepository mergeRepository, IgmRepository igmRepos
         this.igmRepository = igmRepository;
     }
 
-    public void addMergeIgmEvent(String process, String businessProcess, ZonedDateTime date, String tso, IgmStatus status, UUID networkUuid) {
+    public void addMergeIgmEvent(String process, String businessProcess, ZonedDateTime date, String tso, IgmStatus status, UUID networkUuid, UUID caseUuid) {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/MergeEventService.java:54]: why this change ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java [hunk 9]
```diff
@@ -45,21 +68,121 @@ public NetworkConversionService(RestTemplate restTemplate) {
         this.networkConversionServerRest = restTemplate;
     }
 
-    public ExportNetworkInfos exportMerge(List<UUID> networksIds, String format, String baseFileName) {
-        UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromPath(DELIMITER + NETWORK_CONVERSION_API_VERSION + "/networks/{networkUuid}/export/{format}");
+    public void setCaseFetcherService(CaseFetcherService caseFetcherService) {
+        this.caseFetcherService = caseFetcherService;
+    }
+
+    public void setCgmesBoundaryService(CgmesBoundaryService cgmesBoundaryService) {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:75]: we don't have setters for the other dependencies, why add this one ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java [hunk 10]
```diff
@@ -45,21 +68,121 @@ public NetworkConversionService(RestTemplate restTemplate) {
         this.networkConversionServerRest = restTemplate;
     }
 
-    public ExportNetworkInfos exportMerge(List<UUID> networksIds, String format, String baseFileName) {
-        UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromPath(DELIMITER + NETWORK_CONVERSION_API_VERSION + "/networks/{networkUuid}/export/{format}");
+    public void setCaseFetcherService(CaseFetcherService caseFetcherService) {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:71]: we don't have setters for the other dependencies, why add this one ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java [hunk 11]
```diff
@@ -45,21 +68,121 @@ public NetworkConversionService(RestTemplate restTemplate) {
...
+
+            //Add boundary files
+            cgmesProfiles.addAll(getBoundaries());
+
+            //Zip files
+            ByteArrayOutputStream baosZip = createZipFile(cgmesProfiles);
+
+            return new FileInfos(baseFileName.concat(UNDERSCORE + FILE_VERSION + ZIP), baosZip.toByteArray());
+        } else {
+            UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromPath(DELIMITER + NETWORK_CONVERSION_API_VERSION + "/networks/{networkUuid}/export/{format}");
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:101]: don't use UriComponentsBuilder, instead create a template that has "networkUuid={n1}&networkUuid={n2}..." and pass the networkUuids as a variable to RestTemplate.exchange()

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java [hunk 12]
```diff
@@ -45,21 +68,121 @@ public NetworkConversionService(RestTemplate restTemplate) {
...
+            for (int i = 1; i < networkUuids.size(); ++i) {
+                uriBuilder = uriBuilder.queryParam("networkUuid", networkUuids.get(i).toString());
+            }
+            String uri = uriBuilder.build().toUriString();
+            ResponseEntity<byte[]> responseEntity = networkConversionServerRest.exchange(uri, HttpMethod.GET, HttpEntity.EMPTY, new ParameterizedTypeReference<byte[]>() { }, networkUuids.get(0).toString(), format);
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:106]: should we avoid loading the whole thing in memory N

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java (hunk 11 above)
Comment (jonenst) [reply src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:101]: Ah this is from the previous code .. You can leave it then, we will fix it later

File: src/main/java/org/gridsuite/merge/orchestrator/server/MergeOrchestratorService.java [hunk 13]
```diff
@@ -186,13 +189,12 @@ void merge(ProcessConfig processConfig, ZonedDateTime dateTime, String date, Str
         }
     }
 
-    private List<UUID> findNetworkUuidsOfValidatedIgms(ZonedDateTime dateTime, String process) {
+    private List<IgmEntity> findNetworkUuidsOfValidatedIgms(ZonedDateTime dateTime, String process) {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/merge/orchestrator/server/MergeOrchestratorService.java:192]: The method was named findUuids and used to return UUIDs. If you change the returned value, you should also rename the method

File: src/main/java/org/gridsuite/merge/orchestrator/server/MergeEventService.java (hunk 8 above)
Comment (jonenst) [reply src/main/java/org/gridsuite/merge/orchestrator/server/MergeEventService.java:54]: Ah ok we store the caseUuid to be able to redownload the original case when we export.

**Comment by jonenst** [comment]
I have doubts about the design:
- the merge-orchestrator now always stores the original caseUuid (in case we want to export)
- for cgmes exports, it redownloads the original cases from the case-server and downloads the SV profile export from network-conversion-server (which uses the network-store), and manually unzips everything excluding the original SV profiles (and also BD files in the original case) and adds the export SV profile and the last boundary files and rezips the whole thing  
- for other exports, it just delegates to network-conversion-server

is this what you had in mind @geofjamg ?

File: src/main/java/org/gridsuite/merge/orchestrator/server/CgmesBoundaryService.java (hunk 5 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/CgmesBoundaryService.java:60]: done

File: src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java (hunk 4 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java:110]: done

File: src/main/java/org/gridsuite/merge/orchestrator/server/MergeOrchestratorService.java (hunk 13 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/MergeOrchestratorService.java:192]: done

File: src/main/java/org/gridsuite/merge/orchestrator/server/MergeEventService.java (hunk 8 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/MergeEventService.java:54]: Yes

**Comment by geofjamg** [comment]
> I have doubts about the design:
> - the merge-orchestrator now always stores the original caseUuid (in case we want to export)
> - for cgmes exports, it redownloads the original cases from the case-server and downloads the SV profile export from network-conversion-server (which uses the network-store), and manually unzips everything excluding the original SV profiles (and also BD files in the original case) and adds the export SV profile and the last boundary files and rezips the whole thing  
> - for other exports, it just delegates to network-conversion-server
> 
> is this what you had in mind @geofjamg ?

Yes

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java (hunk 10 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:71]: Done

File: src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java (hunk 9 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/NetworkConversionService.java:75]: Done

File: src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java (hunk 6 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java:114]: Removed. This was useless.

**Review by jonenst — APPROVED** [review]

File: src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java (hunk 7 above)
Comment (jonenst) [reply src/main/java/org/gridsuite/merge/orchestrator/server/CaseFetcherService.java:102]: We decided it was ok to have zips in memory, just not their unzipped content
